qa/rpc-tests/wallet_listreceived.py

```diff
@@ -1,98 +0,0 @@
-        self.generate_and_sync(height+1)
-        taddr = self.nodes[1].getnewaddress()
-        zaddr1 = self.nodes[1].z_getnewaddress(release)
-
-        self.nodes[0].sendtoaddress(taddr, 2.0)
-        self.generate_and_sync(height+2)
-
-        # Send 1 ZEC to zaddr1
-        opid = self.nodes[1].z_sendmany(taddr,
-            [{'address': zaddr1, 'amount': 1, 'memo': my_memo}])
-        txid = wait_and_assert_operationid_status(self.nodes[1], opid)
-        self.sync_all()
-        r = self.nodes[1].z_listreceivedbyaddress(zaddr1)
-        assert_equal(0, len(r), "Should have received no confirmed note")
-
-        # No confirmation required, one note should be present
-        r = self.nodes[1].z_listreceivedbyaddress(zaddr1, 0)
-        assert_equal(1, len(r), "Should have received one (unconfirmed) note")
-        assert_equal(txid, r[0]['txid'])
-        assert_equal(1, r[0]['amount'])
-        assert_false(r[0]['change'], "Note should not be change")
-        assert_equal(my_memo, r[0]['memo'])
-
-        # Confirm transaction (1 ZEC from taddr to zaddr1)
-        self.generate_and_sync(height+3)
-
-        # Require one confirmation, note should be present
-        assert_equal(r, self.nodes[1].z_listreceivedbyaddress(zaddr1))
-
-        # Generate some change by sending part of zaddr1 to zaddr2
-        zaddr2 = self.nodes[1].z_getnewaddress(release)
-        opid = self.nodes[1].z_sendmany(zaddr1,
-            [{'address': zaddr2, 'amount': 0.6, 'memo': my_memo}])
-        txid = wait_and_assert_operationid_status(self.nodes[1], opid)
-        self.sync_all()
-        self.generate_and_sync(height+4)
-
-        # zaddr1 should have a note with change
-        r = self.nodes[1].z_listreceivedbyaddress(zaddr1, 0)
-        r = sorted(r, key = lambda received: received['amount'])
-        assert_equal(2, len(r), "zaddr1 Should have received 2 notes")
-
-        assert_equal(txid, r[0]['txid'])
-        assert_equal(Decimal('0.4')-fee, r[0]['amount'])
-        assert_true(r[0]['change'], "Note valued at (0.4-fee) should be change")
-        assert_equal(expected_memo, r[0]['memo'])
-
-        # The old note still exists (it's immutable), even though it is spent
-        assert_equal(Decimal('1.0'), r[1]['amount'])
-        assert_false(r[1]['change'], "Note valued at 1.0 should not be change")
-        assert_equal(expected_memo, r[0]['memo'])
-
-        # zaddr2 should not have change
-        r = self.nodes[1].z_listreceivedbyaddress(zaddr2, 0)
-        r = sorted(r, key = lambda received: received['amount'])
-        assert_equal(1, len(r), "zaddr2 Should have received 1 notes")
-        assert_equal(txid, r[0]['txid'])
-        assert_equal(Decimal('0.6'), r[0]['amount'])
-        assert_false(r[0]['change'], "Note valued at 0.6 should not be change")
-        assert_equal(my_memo, r[0]['memo'])
-
-    def run_test(self):
-        self.run_test_release('sprout', no_memo, 200)
-        self.run_test_release('sapling', no_memo, 204)
```

Comment:
I liked this idea because it made it clear and mandated that sprout and sapling work the same way. I wish we could have separate Sprout and Sapling test files that used this shared logic (separate tests with different parameters in `setup_nodes` would negate the need for https://github.com/zcash/zcash/commit/18f8abb62de00318c69e2a270548e2bdbb554f79). I am not sure if not doing it this way ends up causing duplication in the new version, but it seems likely.

Reply:
Hmm.. I liked it too (that's what I wrote it!), but I didn't structure this updated `wallet_listnotes.py` test this way because I also agree with one of your earlier suggestions that we should be testing more Sapling and less Sprout. `wallet_listreceived.py` tests Sprout and Sapling equally. Since Sprout proofs are much slower, the total run time is more than doubled. (I just timed it, testing both takes 90 seconds, testing only Sapling takes 19 seconds, most of that setup.) Having separate Sprout and Sapling test files wouldn't change that runtime difference (actually would make it more pronounced, since setup would be done twice).

This updated `wallet_listnotes.py` tests Sprout a little and Sapling a lot more. 

It seems like maybe what we should do is have separate Sprout and Sapling test files, calling common (or nearly common; there would have to be a small number of release-specific code, such as I had to do here with the memo) test code, and only run the Sapling tests by default (by Homu) but run all the tests sometimes (not sure what would trigger "sometimes" other than manually). It doesn't feel productive to test RPCs when Sprout (or Overwinter) is active, since they're not run that way once Sapling activates. But I do like what you say about this guaranteeing that Sprout and Sapling work the same way (except as dictated by explicit conditionals).

In any event, restructuring the tests (if we decide to do that) should probably be a separate ticket and PR.